File: voice-agent/voice_agent/realtime_lifecycle.py
"""Realtime session lifecycle + cost guardrails for multi-party rooms (#439).

Part of the Hybrid Realtime Voice epic (#440, Option B: memql is the
director, OpenAI ``gpt-realtime`` is the fast voice executor). This module
bounds a realtime speech-to-speech session so that an always-on model
listening to a polyphon room (up to 5 humans + 1 assistant) cannot burn
audio tokens unbounded. It is the #439 seam tagged in
``realtime_executor.build_realtime_executor``; the executor stays a pure
model factory and this module owns *when* a session lives, mutes, and dies.

Lifecycle decision (acceptance criterion 1)
-------------------------------------------
The issue lays out three lifecycle options:

* (a) session always warm but input-muted until the conductor engages;
* (b) spin up on first engagement, idle-timeout down;
* (c) a hybrid.

The chosen default is **(a) warm-but-muted**, because it composes directly
with the #432 conductor gate. The realtime model is already built with
``turn_detection=None`` (docs/voice/432-conductor-response-gate.md section
2.2, option A): the model never runs input VAD, never commits the input
buffer, and never self-creates a response. memql's conductor is the sole
driver of ``response.create``. That is *exactly* a warm-but-muted posture
already -- a session that has been built but not yet engaged sits with the
model loaded (low warm latency on the first real turn) while consuming no
audio-token cost for generated responses. This module makes the standing
cost of that warm session bounded rather than open-ended:

* **Empty-room teardown** -- when the last human leaves, the warm session
  is torn down immediately. A room with no humans has nobody to talk to;
  keeping the model warm for an empty room is pure waste (acceptance
  criterion 3: "no always-on full-audio session in a multi-listener room
  unless justified").
* **Idle teardown** -- if the conductor never engages the warm session
  within ``realtime_idle_timeout_sec``, it is torn down. A room where the
  assistant is only ever a listener should not hold a warm realtime
  session forever.
* **Max-duration teardown** -- a hard wall-clock backstop
  (``realtime_max_session_sec``) regardless of activity.
* **Token-budget teardown (acceptance criterion 2)** -- the running
  audio-token total is metered per response; crossing
  ``realtime_max_audio_tokens`` tears the session down and signals the
  caller to degrade to the cascade.

Degrade-to-cascade
------------------
Every teardown carries a :class:`TeardownReason`. When the reason is a
cost/idle/duration guardrail (not a normal room close), the lifecycle
exposes ``should_degrade_to_cascade`` so the caller can rebuild the
session on the proven cascade path -- the same "degrade to the cascade
when exceeded" posture the executor selection already uses on a realtime
build failure (``realtime_executor.select_voice_executor``). This module
does not itself rebuild the session; it decides *that* teardown should
happen and *why*, and invokes an injected ``stop`` callback. Wiring the
rebuild is the caller's job in ``main.py``.

Kill-switch composition
-----------------------
The platform-wide computer-use kill switch and per-session controls
(CLAUDE.md "Workers" / "CoPresent Control") already gate agent action at
the memql layer. This module's teardown path is the voice-side analog: a
single ``request_teardown(reason)`` entry point that any external control
signal (a kill-switch push, an operator stop, a session-end envelope) can
call to bring the realtime session down deterministically. The conductor
engaging / a response completing feed the activity clock; nothing here
reaches back into the model except via the injected ``stop`` callback, so
the module stays testable without a live realtime connection.

Out of scope (other epic issues, seams left untouched):

* tool bridge -> #435
* persona / grounding -> #436
* output capture -> #437
* avatar drive -> #438
"""

from __future__ import annotations

import asyncio
import logging
import time
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from enum import Enum

from voice_agent.config import Config

logger = logging.getLogger(__name__)


class TeardownReason(str, Enum):
    """Why a realtime session was (or is being) torn down.

    The ``*_GUARDRAIL`` / ``EMPTY_ROOM`` reasons are cost-driven and mark a
    session that should degrade to the cascade. ``ROOM_CLOSED`` is a normal
    end-of-call teardown that does NOT degrade -- the room is simply over.
    """

    #: The last human left the room. Cost-driven; degrade to cascade if the
    #: room re-populates (the caller decides whether a re-join rebuilds).
    EMPTY_ROOM = "empty_room"
    #: No conductor engagement within the idle window. Cost-driven.
    IDLE_TIMEOUT = "idle_timeout"
    #: Hard wall-clock cap reached. Cost-driven backstop.
    MAX_DURATION = "max_duration"
    #: Per-session audio-token budget exhausted. The cost guardrail.
    TOKEN_BUDGET = "token_budget"
    #: An external control signal (kill switch / operator stop / session
    #: end) requested teardown. Treated as a normal stop, not a degrade.
    EXTERNAL = "external"
    #: The room/call ended normally. Not cost-driven; no cascade degrade.
    ROOM_CLOSED = "room_closed"

    @property
    def is_cost_guardrail(self) -> bool:
        """True when this teardown is a cost/idle guardrail trip.

        These are the reasons the voice path should consider degrading to
        the cascade rather than simply ending: the realtime session was
        stopped to bound cost, not because the call is over.
        """
        return self in (
            TeardownReason.EMPTY_ROOM,
            TeardownReason.IDLE_TIMEOUT,
            TeardownReason.MAX_DURATION,
            TeardownReason.TOKEN_BUDGET,
        )


@dataclass(frozen=True)
class RealtimeBudget:
    """Resolved cost guardrails for one realtime session.

    Built from :class:`Config` via :meth:`from_config`. A ceiling of ``<= 0``
    means that particular guardrail is disabled; the others still apply. All
    durations are seconds; the token budget is total audio tokens
    (input + output) for the session.
    """

    idle_timeout_sec: int
    max_session_sec: int
    max_audio_tokens: int

    @classmethod
    def from_config(cls, cfg: Config) -> RealtimeBudget:
        return cls(
            idle_timeout_sec=cfg.realtime_idle_timeout_sec,
            max_session_sec=cfg.realtime_max_session_sec,
            max_audio_tokens=cfg.realtime_max_audio_tokens,
        )

    @property
    def idle_enabled(self) -> bool:
        return self.idle_timeout_sec > 0

    @property
    def duration_enabled(self) -> bool:
        return self.max_session_sec > 0

    @property
    def tokens_enabled(self) -> bool:
        return self.max_audio_tokens > 0


# A coroutine that actually stops the realtime session. Injected so the
# lifecycle is testable without a live model: in main.py this awaits the
# AgentSession teardown; in tests it records the call.
StopCallback = Callable[[TeardownReason], Awaitable[None]]

# A monotonic clock source, injectable so tests can advance time without
# sleeping. Defaults to time.monotonic.
Clock = Callable[[], float]


@dataclass
class _State:
    """Mutable lifecycle state, guarded by the manager's lock."""

    started_at: float
    last_engaged_at: float
    human_count: int = 0
    audio_tokens_used: int = 0
    torn_down: bool = False
    teardown_reason: TeardownReason | None = None


class RealtimeSessionLifecycle:
    """Owns the lifecycle of one realtime session in a multi-party room.

    Construction does NOT start the session -- the caller builds the model
    (``realtime_executor.build_realtime_executor``) and the
    ``AgentSession`` itself. This object is then started with :meth:`start`,
    which begins the warm session in the muted posture and arms the idle /
    duration watchdog. The caller feeds it room + activity events:

    * :meth:`note_human_joined` / :meth:`note_human_left` -- room
      participant lifecycle. The last human leaving triggers an
      empty-room teardown.
    * :meth:`note_engaged` -- the conductor drove a ``response.create``;
      resets the idle clock.
    * :meth:`note_audio_tokens` -- usage reported for a completed
      response; trips the token-budget guardrail when the running total
      crosses the ceiling.
    * :meth:`request_teardown` -- external stop (kill switch / operator /
      session end).

    Teardown is idempotent and single-shot: the first reason wins, the
    injected ``stop`` callback runs exactly once, and subsequent events are
    no-ops. After teardown, :attr:`should_degrade_to_cascade` tells the
    caller whether to rebuild on the cascade.
    """

    def __init__(
        self,
        *,
        budget: RealtimeBudget,
        stop: StopCallback,
        space_id: str = "",
        clock: Clock = time.monotonic,
        watchdog_interval_sec: float = 5.0,
    ) -> None:
        self._budget = budget
        self._stop = stop
        self._space_id = space_id
        self._clock = clock
        self._watchdog_interval = max(0.1, watchdog_interval_sec)
        self._lock = asyncio.Lock()
        self._state: _State | None = None
        self._watchdog: asyncio.Task[None] | None = None

    # -- lifecycle ---------------------------------------------------------

    async def start(self, *, human_count: int = 0) -> None:
        """Begin the warm session and arm the idle / duration watchdog.

        ``human_count`` seeds the room population (the entrypoint waits for
        at least one participant before building the session, so this is
        normally >= 1). Starting with zero humans is allowed but arms the
        empty-room teardown on the next watchdog tick.
        """
        now = self._clock()
        self._state = _State(
            started_at=now,
            last_engaged_at=now,
            human_count=human_count,
        )
        logger.info(
            "realtime lifecycle started space=%s humans=%d idle_timeout=%ds "
            "max_session=%ds token_budget=%d (warm-but-muted, #439)",
            self._space_id,
            human_count,
            self._budget.idle_timeout_sec,
            self._budget.max_session_sec,
            self._budget.max_audio_tokens,
        )
        if self._budget.idle_enabled or self._budget.duration_enabled:
            self._watchdog = asyncio.create_task(self._watchdog_loop())

    async def aclose(self) -> None:
        """Stop the watchdog without tearing the session down via a reason.

        For caller-side cleanup paths (the entrypoint ``finally`` block)
        that already own session teardown and just need the watchdog task
        cancelled. Does not invoke the ``stop`` callback.
        """
        await self._cancel_watchdog()

    # -- event inputs ------------------------------------------------------

    async def note_human_joined(self) -> None:
        """A human participant joined. Bumps the room population."""
        async with self._lock:
            if self._state is None or self._state.torn_down:
                return
            self._state.human_count += 1
            logger.debug(
                "realtime lifecycle human joined space=%s humans=%d",
                self._space_id,
                self._state.human_count,
            )

    async def note_human_left(self) -> None:
        """A human participant left.

        When the population reaches zero the session is torn down with
        :attr:`TeardownReason.EMPTY_ROOM` -- a room with no humans has
        nobody for the assistant to talk to, so the warm session is pure
        cost.
        """
        async with self._lock:
            if self._state is None or self._state.torn_down:
                return
            self._state.human_count = max(0, self._state.human_count - 1)
            logger.debug(
                "realtime lifecycle human left space=%s humans=%d",
                self._space_id,
                self._state.human_count,
            )
            if self._state.human_count == 0:
                await self._teardown_locked(TeardownReason.EMPTY_ROOM)

    async def note_engaged(self) -> None:
        """The conductor engaged the session (drove a ``response.create``).

        Resets the idle clock: an engaged session is doing useful work and
        should not be idle-reaped out from under an active conversation.
        """
        async with self._lock:
            if self._state is None or self._state.torn_down:
                return
            self._state.last_engaged_at = self._clock()
            logger.debug(
                "realtime lifecycle engaged space=%s -- idle clock reset",
                self._space_id,
            )

    async def note_audio_tokens(self, tokens: int) -> None:
        """Record audio-token usage reported for a completed response.

        ``tokens`` is the input+output audio-token count for one response
        (the realtime model reports usage per response). When the running
        session total crosses ``max_audio_tokens`` the session is torn down
        with :attr:`TeardownReason.TOKEN_BUDGET` and the caller should
        degrade to the cascade. Negative counts are clamped to zero.
        """
        async with self._lock:
            if self._state is None or self._state.torn_down:
                return
            self._state.audio_tokens_used += max(0, tokens)
            if (
                self._budget.tokens_enabled
                and self._state.audio_tokens_used >= self._budget.max_audio_tokens
            ):
                logger.warning(
                    "realtime audio-token budget exhausted space=%s used=%d "
                    "budget=%d -- tearing down + degrading to cascade (#439)",
                    self._space_id,
                    self._state.audio_tokens_used,
                    self._budget.max_audio_tokens,
                )
                await self._teardown_locked(TeardownReason.TOKEN_BUDGET)

    async def request_teardown(self, reason: TeardownReason = TeardownReason.EXTERNAL) -> None:
        """Externally request teardown (kill switch / operator / session end).

        Idempotent: the first teardown reason wins.
        """
        async with self._lock:
            await self._teardown_locked(reason)

    # -- watchdog ----------------------------------------------------------

    async def _watchdog_loop(self) -> None:
        """Poll for idle / max-duration expiry on a fixed interval.

        Polling (rather than one-shot timers) keeps the logic simple and
        robust to engagement events resetting the idle clock: each tick
        re-reads the current deltas against the live budget. The interval
        is small relative to the guardrail windows so the teardown fires
        within one tick of the deadline.
        """
        try:
            while True:
                await asyncio.sleep(self._watchdog_interval)
                async with self._lock:
                    if self._state is None or self._state.torn_down:
                        return
                    now = self._clock()
                    if (
                        self._budget.duration_enabled
                        and now - self._state.started_at >= self._budget.max_session_sec
                    ):
                        logger.warning(
                            "realtime max session duration reached space=%s "
                            "elapsed=%.0fs cap=%ds -- tearing down (#439)",
                            self._space_id,
                            now - self._state.started_at,
                            self._budget.max_session_sec,
                        )
                        await self._teardown_locked(TeardownReason.MAX_DURATION)
                        return
                    if (
                        self._budget.idle_enabled
                        and now - self._state.last_engaged_at
                        >= self._budget.idle_timeout_sec
                    ):
                        logger.warning(
                            "realtime session idle space=%s idle_for=%.0fs "
                            "timeout=%ds -- tearing down (#439)",
                            self._space_id,
                            now - self._state.last_engaged_at,
                            self._budget.idle_timeout_sec,
                        )
                        await self._teardown_locked(TeardownReason.IDLE_TIMEOUT)
                        return
        except asyncio.CancelledError:  # pragma: no cover - cancellation path
            raise

    async def _cancel_watchdog(self) -> None:
        task = self._watchdog
        self._watchdog = None
        if task is None:
            return
        task.cancel()
        try:
            await task
        except asyncio.CancelledError:
            pass

    # -- teardown ----------------------------------------------------------

    async def _teardown_locked(self, reason: TeardownReason) -> None:
        """Single-shot teardown. Caller must hold ``self._lock``.

        Records the reason, invokes the injected ``stop`` callback exactly
        once, and cancels the watchdog. A failure in the stop callback is
        logged but does not unwind the torn-down state -- the session is
        considered down regardless so no event re-fires teardown.
        """
        if self._state is None or self._state.torn_down:
            return
        self._state.torn_down = True
        self._state.teardown_reason = reason
        logger.info(
            "realtime lifecycle teardown space=%s reason=%s degrade_to_cascade=%s "
            "tokens_used=%d (#439)",
            self._space_id,
            reason.value,
            reason.is_cost_guardrail,
            self._state.audio_tokens_used,
        )
        try:
            await self._stop(reason)
        except Exception:  # noqa: BLE001 -- stop failure must not wedge teardown
            logger.exception(
                "realtime lifecycle stop callback failed space=%s reason=%s",
                self._space_id,
                reason.value,
            )
        # Cancel the watchdog OUTSIDE the lock would risk re-entrancy with
        # _watchdog_loop awaiting the lock; instead null the handle here and
        # let aclose / the loop's own torn_down check stop it. The loop
        # returns on the next tick when it sees torn_down=True.
        if self._watchdog is not None and self._watchdog is not asyncio.current_task():
            self._watchdog.cancel()
            self._watchdog = None

    # -- introspection -----------------------------------------------------

    @property
    def is_torn_down(self) -> bool:
        return self._state is not None and self._state.torn_down

    @property
    def teardown_reason(self) -> TeardownReason | None:
        return self._state.teardown_reason if self._state is not None else None

    @property
    def should_degrade_to_cascade(self) -> bool:
        """True when the session was torn down by a cost guardrail.

        The caller reads this after teardown to decide whether to rebuild
        the voice path on the cascade (cost guardrail trip) or simply let
        the call end (normal room close / external stop).
        """
        reason = self.teardown_reason
        return reason is not None and reason.is_cost_guardrail

    @property
    def audio_tokens_used(self) -> int:
        return self._state.audio_tokens_used if self._state is not None else 0

    @property
    def human_count(self) -> int:
        return self._state.human_count if self._state is not None else 0


def build_realtime_lifecycle(
    cfg: Config,
    *,
    stop: StopCallback,
    space_id: str = "",
    clock: Clock = time.monotonic,
) -> RealtimeSessionLifecycle:
    """Construct a :class:`RealtimeSessionLifecycle` from runtime config.

    Convenience factory mirroring ``build_realtime_executor`` so ``main.py``
    wires the lifecycle in one call alongside the executor. The caller still
    invokes :meth:`RealtimeSessionLifecycle.start` once the session is up.
    """
    return RealtimeSessionLifecycle(
        budget=RealtimeBudget.from_config(cfg),
        stop=stop,
        space_id=space_id,
        clock=clock,
    )

File: integrations/voice/agent/realtime_lifecycle.go
package agent

import (
	"log/slog"
	"sync"
	"time"
)

// realtime_lifecycle.go is the Go analog of the RealtimeSessionLifecycle half of
// the Python voice-agent's realtime_lifecycle.py (#439 -> #459). It bounds one
// realtime speech-to-speech session in a multi-party room so that an always-on
// model listening to a polyphon room (up to 5 humans + 1 assistant) cannot burn
// audio tokens unbounded. It owns *when* a warm session lives, mutes, and dies;
// the executor (realtime_executor.go) stays the turn-taking / wire driver.
//
// It is PURE Go and CGO-free -- timers, counters, a mutex, and an injected stop
// callback. The watchdog is the Go analog of the Python asyncio poll loop: a
// single goroutine driven by a time.Ticker that re-reads idle / max-duration
// deltas against the live budget each tick, so an engagement event resetting the
// idle clock is honoured on the next tick. For deterministic unit tests the
// clock is injectable (a fake monotonic source) and the guardrail check is
// reachable directly (checkExpiry) without sleeping; the goroutine is only the
// production driver.
//
// Lifecycle decision (parity with the Python module): warm-but-muted (option a).
// The realtime session is built with turn_detection:null and the conductor is
// the sole driver of response.create (see realtime_executor.go / #457), so a
// built-but-not-yet-engaged session already sits warm and muted. This module
// makes the standing cost bounded:
//
//   - Empty-room teardown -- when the last human leaves, the warm session is torn
//     down immediately (ReasonEmptyRoom).
//   - Idle teardown -- if the conductor never engages within IdleTimeoutSec, the
//     session is torn down (ReasonIdleTimeout).
//   - Max-duration teardown -- a hard wall-clock backstop (ReasonMaxDuration).
//   - Token-budget teardown -- the running audio-token total crossing
//     MaxAudioTokens tears the session down (ReasonTokenBudget) and signals the
//     caller to degrade to the cascade.
//
// Degrade-to-cascade: every teardown carries a TeardownReason. When the reason
// is a cost guardrail (TeardownReason.IsCostGuardrail), ShouldDegradeToCascade
// is true so the caller can rebuild on the proven cascade path -- the same
// posture executor_select.go uses on a realtime build failure. This module does
// not itself rebuild; it decides THAT teardown should happen and WHY, and invokes
// the injected stop callback exactly once.

// realtimeStopFunc actually stops the realtime session. Injected so the
// lifecycle is testable without a live model: in the executor wiring it closes
// the session; in tests it records the call. A non-nil error is logged but does
// NOT unwind the torn-down state -- the session is considered down regardless so
// no later event re-fires teardown. Mirrors the Python StopCallback.
type realtimeStopFunc func(reason TeardownReason) error

// realtimeClock is a monotonic clock source, injectable so tests advance time
// without sleeping. Defaults to time.Now in NewRealtimeLifecycle. Mirrors the
// Python Clock seam.
type realtimeClock func() time.Time

// defaultWatchdogInterval is how often the production watchdog goroutine
// re-evaluates idle / max-duration expiry. Small relative to the guardrail
// windows (idle default 300s, max default 1800s) so teardown fires within one
// tick of the deadline.
const defaultWatchdogInterval = 5 * time.Second

// RealtimeSessionLifecycle owns the lifecycle of one realtime session in a
// multi-party room. Construction does NOT start the session -- the caller builds
// the model + session and the executor, then calls Start, which records the
// start time, seeds the room population, and (when an idle / duration guardrail
// is enabled) arms the watchdog goroutine. The caller feeds it room + activity
// events:
//
//   - NoteHumanJoined / NoteHumanLeft -- room participant lifecycle. The last
//     human leaving triggers an empty-room teardown.
//   - NoteEngaged -- the conductor drove a response.create; resets the idle clock.
//   - NoteAudioTokens -- usage reported for a completed response; trips the
//     token-budget guardrail when the running total crosses the ceiling.
//   - RequestTeardown -- external stop (kill switch / operator / session end).
//
// Teardown is idempotent and single-shot: the first reason wins, the injected
// stop callback runs exactly once, and subsequent events are no-ops. After
// teardown, ShouldDegradeToCascade tells the caller whether to rebuild on the
// cascade. Mirrors realtime_lifecycle.py::RealtimeSessionLifecycle.
type RealtimeSessionLifecycle struct {
	budget   RealtimeBudget
	stop     realtimeStopFunc
	spaceID  string
	clock    realtimeClock
	interval time.Duration
	logger   *slog.Logger

	// tearDownOnEmptyRoom controls whether the population reaching zero hard-tears
	// the warm session down immediately (ReasonEmptyRoom). Default OFF (#1449):
	// a mic toggle in the SPA surfaces as a transient LiveKit participant
	// disconnect+reconnect, so an instant empty-room teardown closes the
	// gpt-realtime session out from under a user who is about to re-enable their
	// mic -- then re-engage updates a dead handle and voice never recovers.
	// With this off the session stays warm-but-muted at zero humans (the #459
	// posture); the room-level idle watchdog (#1378, 60s grace counting real
	// participants) owns genuine departures, and the idle / max-duration / token
	// guardrails still bound a warm-but-idle session's cost. When ON, the legacy
	// reflexive teardown applies.
	tearDownOnEmptyRoom bool

	mu             sync.Mutex
	started        bool
	startedAt      time.Time
	lastEngagedAt  time.Time
	humanCount     int
	audioTokens    int
	tornDown       bool
	teardownReason TeardownReason

	// watchdog stop signalling. stopCh is closed once (single watchdog).
	stopCh   chan struct{}
	stopOnce sync.Once
}

// realtimeLifecycleOption configures a RealtimeSessionLifecycle at construction.
// The clock and watchdog interval are injectable for deterministic tests.
type realtimeLifecycleOption func(*RealtimeSessionLifecycle)

// withLifecycleClock injects a clock source (tests advance a fake monotonic
// clock without sleeping). Defaults to time.Now.
func withLifecycleClock(clock realtimeClock) realtimeLifecycleOption {
	return func(l *RealtimeSessionLifecycle) {
		if clock != nil {
			l.clock = clock
		}
	}
}

// withWatchdogInterval overrides the watchdog poll interval (tests use a short
// interval; production uses defaultWatchdogInterval).
func withWatchdogInterval(d time.Duration) realtimeLifecycleOption {
	return func(l *RealtimeSessionLifecycle) {
		if d > 0 {
			l.interval = d
		}
	}
}

// withLifecyclePartitionID stamps the space id on the lifecycle's log lines.
func withLifecyclePartitionID(spaceID string) realtimeLifecycleOption {
	return func(l *RealtimeSessionLifecycle) { l.spaceID = spaceID }
}

// withEmptyRoomTeardown opts the lifecycle back into the legacy reflexive
// empty-room hard-teardown when the population reaches zero (#1449 default is
// off: warm-but-muted, the idle watchdog + idle guardrail govern an empty room).
func withEmptyRoomTeardown(enabled bool) realtimeLifecycleOption {
	return func(l *RealtimeSessionLifecycle) { l.tearDownOnEmptyRoom = enabled }
}

// NewRealtimeLifecycle builds a lifecycle from a resolved budget and an injected
// stop callback. The caller still invokes Start once the session is up. Mirrors
// realtime_lifecycle.py::build_realtime_lifecycle.
func NewRealtimeLifecycle(
	budget RealtimeBudget,
	stop realtimeStopFunc,
	logger *slog.Logger,
	opts ...realtimeLifecycleOption,
) *RealtimeSessionLifecycle {
	l := &RealtimeSessionLifecycle{
		budget:   budget,
		stop:     stop,
		clock:    time.Now,
		interval: defaultWatchdogInterval,
		logger:   logger,
		stopCh:   make(chan struct{}),
	}
	for _, opt := range opts {
		opt(l)
	}
	return l
}

// Start begins the warm session's bookkeeping and arms the idle / duration
// watchdog. humanCount seeds the room population (the room layer waits for at
// least one participant before building the session, so this is normally >= 1).
// Starting with zero humans is allowed but arms the empty-room teardown on the
// next watchdog tick. Idempotent: a second Start is a no-op.
func (l *RealtimeSessionLifecycle) Start(humanCount int) {
	l.mu.Lock()
	if l.started || l.tornDown {
		l.mu.Unlock()
		return
	}
	now := l.clock()
	l.started = true
	l.startedAt = now
	l.lastEngagedAt = now
	if humanCount > 0 {
		l.humanCount = humanCount
	}
	armWatchdog := l.budget.IdleEnabled() || l.budget.DurationEnabled()
	l.mu.Unlock()

	if l.logger != nil {
		l.logger.Info("voice-agent realtime lifecycle started (warm-but-muted, #459)",
			"space", l.spaceID,
			"humans", humanCount,
			"idle_timeout_sec", l.budget.IdleTimeoutSec,
			"max_session_sec", l.budget.MaxSessionSec,
			"max_audio_tokens", l.budget.MaxAudioTokens)
	}
	if armWatchdog {
		go l.watchdogLoop()
	}
}

// Close stops the watchdog without invoking a teardown reason. For caller-side
// cleanup paths that already own session teardown and just need the watchdog
// goroutine stopped. Does not invoke the stop callback. Idempotent. Mirrors the
// Python aclose.
func (l *RealtimeSessionLifecycle) Close() {
	l.stopWatchdog()
}

// NoteHumanJoined records a human participant joining (bumps the room
// population). No-op after teardown.
func (l *RealtimeSessionLifecycle) NoteHumanJoined() {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.tornDown {
		return
	}
	l.humanCount++
	if l.logger != nil {
		l.logger.Debug("voice-agent realtime lifecycle human joined",
			"space", l.spaceID, "humans", l.humanCount)
	}
}

// NoteHumanLeft records a human participant leaving. When the population reaches
// zero the session is, by default (#1449), kept warm-but-muted rather than torn
// down: a mic toggle in the SPA surfaces as a transient participant
// disconnect+reconnect, and an instant empty-room teardown closes the
// gpt-realtime session out from under a user who is about to re-enable their mic
// (then the re-engage path updates a dead handle and voice never recovers). The
// room-level idle watchdog (#1378) owns genuine departures, and the idle /
// max-duration / token guardrails still bound the warm-but-idle session's cost.
// Opt back into the reflexive teardown with withEmptyRoomTeardown(true). No-op
// after teardown.
func (l *RealtimeSessionLifecycle) NoteHumanLeft() {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.tornDown {
		return
	}
	if l.humanCount > 0 {
		l.humanCount--
	}
	if l.logger != nil {
		l.logger.Debug("voice-agent realtime lifecycle human left",
			"space", l.spaceID, "humans", l.humanCount,
			"empty_room_teardown", l.tearDownOnEmptyRoom)
	}
	if l.humanCount == 0 && l.tearDownOnEmptyRoom {
		l.teardownLocked(ReasonEmptyRoom)
	}
}

// NoteEngaged records that the conductor engaged the session (drove a
// response.create). Resets the idle clock: an engaged session is doing useful
// work and should not be idle-reaped out from under an active conversation.
// No-op after teardown.
func (l *RealtimeSessionLifecycle) NoteEngaged() {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.tornDown {
		return
	}
	l.lastEngagedAt = l.clock()
	if l.logger != nil {
		l.logger.Debug("voice-agent realtime lifecycle engaged -- idle clock reset",
			"space", l.spaceID)
	}
}

// NoteAudioTokens records audio-token usage reported for a completed response
// (input+output audio tokens, from response.usage on EventResponseDone). When
// the running session total crosses MaxAudioTokens the session is torn down with
// ReasonTokenBudget and the caller should degrade to the cascade. Negative
// counts are clamped to zero. No-op after teardown.
func (l *RealtimeSessionLifecycle) NoteAudioTokens(tokens int) {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.tornDown {
		return
	}
	if tokens > 0 {
		l.audioTokens += tokens
	}
	if l.budget.TokensEnabled() && l.audioTokens >= l.budget.MaxAudioTokens {
		if l.logger != nil {
			l.logger.Warn("voice-agent realtime audio-token budget exhausted -- tearing down + degrading to cascade (#459)",
				"space", l.spaceID,
				"used", l.audioTokens,
				"budget", l.budget.MaxAudioTokens)
		}
		l.teardownLocked(ReasonTokenBudget)
	}
}

// RequestTeardown externally requests teardown (kill switch / operator stop /
// session end). Idempotent: the first teardown reason wins. The Go analog of the
// Python request_teardown entry point.
func (l *RealtimeSessionLifecycle) RequestTeardown(reason TeardownReason) {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.teardownLocked(reason)
}

// watchdogLoop polls for idle / max-duration expiry on the configured interval.
// Polling (rather than one-shot timers) keeps the logic robust to engagement
// events resetting the idle clock: each tick re-reads the current deltas against
// the live budget. The interval is small relative to the guardrail windows so
// teardown fires within one tick of the deadline. Exits on teardown or Close.
func (l *RealtimeSessionLifecycle) watchdogLoop() {
	ticker := time.NewTicker(l.interval)
	defer ticker.Stop()
	for {
		select {
		case <-l.stopCh:
			return
		case <-ticker.C:
			if done := l.checkExpiry(); done {
				return
			}
		}
	}
}

// checkExpiry evaluates the idle / max-duration guardrails against the current
// clock and tears the session down if either has elapsed. It returns true when
// the watchdog should stop (session torn down or already down). Split out from
// the goroutine so tests can drive it deterministically against a fake clock
// without sleeping. Max-duration is checked before idle so a session that has
// both expired reports the wall-clock cap.
func (l *RealtimeSessionLifecycle) checkExpiry() bool {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.tornDown || !l.started {
		return true
	}
	now := l.clock()
	if l.budget.DurationEnabled() && now.Sub(l.startedAt) >= l.durationWindow() {
		if l.logger != nil {
			l.logger.Warn("voice-agent realtime max session duration reached -- tearing down (#459)",
				"space", l.spaceID,
				"elapsed_sec", int(now.Sub(l.startedAt).Seconds()),
				"cap_sec", l.budget.MaxSessionSec)
		}
		l.teardownLocked(ReasonMaxDuration)
		return true
	}
	if l.budget.IdleEnabled() && now.Sub(l.lastEngagedAt) >= l.idleWindow() {
		if l.logger != nil {
			l.logger.Warn("voice-agent realtime session idle -- tearing down (#459)",
				"space", l.spaceID,
				"idle_for_sec", int(now.Sub(l.lastEngagedAt).Seconds()),
				"timeout_sec", l.budget.IdleTimeoutSec)
		}
		l.teardownLocked(ReasonIdleTimeout)
		return true
	}
	return false
}

func (l *RealtimeSessionLifecycle) idleWindow() time.Duration {
	return time.Duration(l.budget.IdleTimeoutSec) * time.Second
}

func (l *RealtimeSessionLifecycle) durationWindow() time.Duration {
	return time.Duration(l.budget.MaxSessionSec) * time.Second
}

// teardownLocked performs the single-shot teardown. The caller must hold l.mu.
// It records the reason, invokes the injected stop callback exactly once, and
// signals the watchdog to exit. A failure in the stop callback is logged but
// does NOT unwind the torn-down state -- the session is considered down
// regardless so no later event re-fires teardown. Mirrors the Python
// _teardown_locked.
func (l *RealtimeSessionLifecycle) teardownLocked(reason TeardownReason) {
	if l.tornDown {
		return
	}
	l.tornDown = true
	l.teardownReason = reason
	if l.logger != nil {
		l.logger.Info("voice-agent realtime lifecycle teardown (#459)",
			"space", l.spaceID,
			"reason", string(reason),
			"degrade_to_cascade", reason.IsCostGuardrail(),
			"tokens_used", l.audioTokens)
	}
	if l.stop != nil {
		if err := l.stop(reason); err != nil && l.logger != nil {
			l.logger.Error("voice-agent realtime lifecycle stop callback failed",
				"space", l.spaceID, "reason", string(reason), "err", err)
		}
	}
	l.signalWatchdogStop()
}

// stopWatchdog signals the watchdog goroutine to exit (Close path). Idempotent.
func (l *RealtimeSessionLifecycle) stopWatchdog() {
	l.signalWatchdogStop()
}

// signalWatchdogStop closes the stop channel exactly once. Safe to call under or
// outside the lock; the sync.Once guards the double-close.
func (l *RealtimeSessionLifecycle) signalWatchdogStop() {
	l.stopOnce.Do(func() { close(l.stopCh) })
}

// IsTornDown reports whether the session has been torn down.
func (l *RealtimeSessionLifecycle) IsTornDown() bool {
	l.mu.Lock()
	defer l.mu.Unlock()
	return l.tornDown
}

// TeardownReason returns the reason the session was torn down, or the empty
// string if it has not been.
func (l *RealtimeSessionLifecycle) TeardownReason() TeardownReason {
	l.mu.Lock()
	defer l.mu.Unlock()
	return l.teardownReason
}

// ShouldDegradeToCascade reports whether the session was torn down by a cost
// guardrail. The caller reads this after teardown to decide whether to rebuild
// the voice path on the cascade (cost guardrail trip) or simply let the call end
// (normal room close / external stop). Mirrors the Python property.
func (l *RealtimeSessionLifecycle) ShouldDegradeToCascade() bool {
	l.mu.Lock()
	defer l.mu.Unlock()
	return l.tornDown && l.teardownReason.IsCostGuardrail()
}

// AudioTokensUsed returns the running audio-token total.
func (l *RealtimeSessionLifecycle) AudioTokensUsed() int {
	l.mu.Lock()
	defer l.mu.Unlock()
	return l.audioTokens
}

// HumanCount returns the current room human population.
func (l *RealtimeSessionLifecycle) HumanCount() int {
	l.mu.Lock()
	defer l.mu.Unlock()
	return l.humanCount
}
